lib/liquid/static_registers.rb

```diff
@@ -0,0 +1,36 @@
+module Liquid
+  class StaticRegisters
+    def initialize(registers = {})
+      @static_registers = registers.is_a?(StaticRegisters) ? registers.static : registers.freeze
+      @registers = {}
+    end
+
+    def []=(key, value)
+      @registers[key] = value
+    end
+
+    def [](key)
+      if @registers.key?(key)
+        @registers[key]
+      else
+        @static_registers[key]
+      end
+    end
+
+    def delete(key)
+      @registers.delete(key)
+    end
+
+    def fetch(key, default = nil)
+      self[key] || default
+    end
+
+    def key?(key)
+      self[key] != nil
```

Comment:
That's incorrect since the key could have been assigned `nil` (in which case `key?` should return true). 

```ruby
@registers.key?(key) || @static_registers.key?(key)
```